File: app/src/main/java/io/mgba/Adapters/Interfaces/SingleViewHolderAdapter.java
package io.mgba.Adapters.Interfaces;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.google.common.base.Function;

import java.util.List;

public abstract class SingleViewHolderAdapter<T> extends BaseAdapter<T> {
    protected final int layout;
    private final Function<View, RecyclerView.ViewHolder> createView;


    public SingleViewHolderAdapter(Context context, RecyclerView recyclerView, int layout, Function<View, RecyclerView.ViewHolder> createView) {
        super(context, recyclerView);
        this.layout = layout;
        this.createView = createView;
    }

    public SingleViewHolderAdapter(List<T> items, RecyclerView recyclerView, Context context, int layout, Function<View, RecyclerView.ViewHolder> createView) {
        super(items, recyclerView, context);
        this.layout = layout;
        this.createView = createView;
    }

    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View inflatedView = LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
        return createView.apply(inflatedView);
    }

}

File: app/src/main/java/io/mgba/adapters/interfaces/SingleViewHolderAdapter.kt
package io.mgba.adapters.interfaces

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView

abstract class SingleViewHolderAdapter<T> : BaseAdapter<T> {
    protected val layout: Int
    private val createView: (View) -> RecyclerView.ViewHolder


    constructor(context: Context, recyclerView: RecyclerView, layout: Int, createView: (View) -> RecyclerView.ViewHolder) : super(context, recyclerView) {
        this.layout = layout
        this.createView = createView
    }

    constructor(items: List<T>, recyclerView: RecyclerView, context: Context, layout: Int, createView: (View) -> RecyclerView.ViewHolder) : super(items, recyclerView, context) {
        this.layout = layout
        this.createView = createView
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val inflatedView = LayoutInflater.from(parent.context).inflate(layout, parent, false)
        return createView.invoke(inflatedView)
    }

}
